third_party/move/move-compiler-v2/src/ast_simplifier.rs

```diff
@@ -0,0 +1,929 @@
+                    modifying = false;
+                },
+                VisitorPosition::PreSequenceValue => {
+                    modifying = modifying_stack.pop().expect("unbalanced visit 6");
+                },
+                _ => {},
+            },
+            Assign(_loc, pat, _) => {
+                if let VisitorPosition::Pre = pos {
+                    // add vars in pat to modified vars
+                    for (_pat_var_id, sym) in pat.vars() {
+                        let current_binding_id = visiting_binding.get(&sym);
+                        trace!(
+                            "Var {} in assignment {:?} is unsafe",
+                            sym.display(env.symbol_pool()),
+                            current_binding_id
+                        );
+                        unsafe_variables.insert((sym, current_binding_id.copied()));
+                    }
+                };
+            },
+        };
+        true
+    });
+    unsafe_variables
+}
+
+/// A function-specific simplifier rewriter.
+struct SimplifierRewriter<'env> {
+    pub env: &'env GlobalEnv,
+    pub func_env: &'env FunctionEnv<'env>,
+
+    pub constant_folder: ConstantFolder<'env>,
+
+    // Guard whether entire subexpressions are eliminated (possibly hiding some warnings).
+    pub eliminate_code: bool,
+
+    // Tracks which definition (`Let` statement `NodeId`) is visible during visit to find modified
+    // local vars.  A use of a symbol which is missing must be a `Parameter`.  This is used only
+    // to determine if a symbol is in `unsafe_variables`.
+    visiting_binding: ScopedMap<Symbol, NodeId>,
+
+    // Unsafe variables are identified by `Symbol` and `Let` statement `NodeId`,
+    // except function parameters, which have no `NodeId` so get `None`.
+    unsafe_variables: BTreeSet<(Symbol, Option<NodeId>)>,
+
+    // Tracks constant values from scope.
+    values: ScopedMap<Symbol, SimpleValue>,
+}
+
+// Representation to record a known value of a variable to
+// allow simplification.  Currently we only track constant values
+// and definitely uninitialzed values (from `let` with no binding).
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+enum SimpleValue {
+    Value(Value),
+    Uninitialized,
+}
+
+impl<'env> SimplifierRewriter<'env> {
+    fn new(env: &'env GlobalEnv, func_env: &'env FunctionEnv, eliminate_code: bool) -> Self {
+        let constant_folder = ConstantFolder::new(env, false);
+        Self {
+            env,
+            func_env,
+            constant_folder,
+            eliminate_code,
+            visiting_binding: ScopedMap::new(),
+            unsafe_variables: BTreeSet::new(),
+            values: ScopedMap::new(),
+        }
+    }
+
+    /// Process a function.
+    pub fn rewrite_function_body(&mut self, exp: Exp) -> Exp {
+        self.unsafe_variables =
+            find_possibly_modified_vars(self.env, self.func_env.get_parameters_ref(), exp.as_ref());
```

Comment:
I'd make this a member function of SimplifierRewriter, also because of evolution of the code. Then you always have all context (fun_env) available for this function. Your call.